twister: Fix tests using required_snippets

nordicjm opened the pull request:
Fixes an issue whereby twister would run tests with a
    required_snippet set and it would apply the snippet during the
    Kconfig filtering stage without all the modules being available


    Adds support for using the Zephyr module snippet_root variable
    in twister for required_snippets

Fixes #63011

rexut commented:
> Adds support for twister automatically checking which modules have snippet directories and including them when searching for required snippets
> 
> Fixes #63011

I agree, it'll fix the test plan strategie to find snippets also in Zephyr modules.

**REVIEW: +1**

nordicjm commented:
@rexut I've updated it, can you give it another try?

rexut commented:
> @rexut I've updated it, can you give it another try?

I do so, but no success. The special `SCOPE snippets` still doesn't exists. I'm also out of focus where in all the cmak modules stuff this scope will be created in an always proper way, if `west build …` will be used instead of `west twister …`.

nordicjm commented:
Maybe that is an issue with the alternative root, as I can't reproduce it just using twister with a modified hello_world. @gmarull any clue on this issue (from https://github.com/zephyrproject-rtos/zephyr/pull/57806) ?

nordicjm committed:
twister: Fix tests using required_snippets

Fixes an issue whereby twister would run tests with a
required_snippet set and it would apply the snippet during the
Kconfig filtering stage without all the modules being available

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
@rexut I think it should now finally be fixed!

tejlmand commented on cmake/modules/snippets.cmake:
```diff
@@ -65,6 +65,15 @@ function(zephyr_process_snippets)
     list(APPEND real_snippet_root ${real_snippet_dir})
   endforeach()
+
+  # Search through all modules for any that have a snippets directory
+  foreach(module_name ${ZEPHYR_MODULE_NAMES})
+    string(TOUPPER "${module_name}" module_name)
+    if(EXISTS "${ZEPHYR_${module_name}_MODULE_DIR}/snippets")
+      list(APPEND real_snippet_root "${ZEPHYR_${module_name}_MODULE_DIR}")
+    endif()
+  endforeach()
```
we should not automatically loop all modules for a snippet root.

The `zephyr/module.yml` file of the module should instead use:
```
build:
  settings:
    snippet_root: <relative-path-to-root>
```
Ref: https://docs.zephyrproject.org/latest/develop/modules.html#build-settings
> snippet_root: Contains additional snippets that are available for use. These snippets must be defined in snippet.yml files underneath the <snippet_root>/snippets folder. 

```suggestion
```

tejlmand commented on doc/releases/release-notes-3.5.rst:
```diff
@@ -128,6 +128,9 @@ Build system and infrastructure
   be used to include a snippet when a test is ran (and exclude any boards from
   running that the snippet cannot be applied to).
 
+* Added support for snippet roots being automatically detected and used if they
+  exist in module directories.
+
```
This is already supported through the use of `zephyr/module.yml`.

```suggestion
```

tejlmand commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -309,6 +310,11 @@ def handle_modules(self):
         modules_meta = parse_modules(ZEPHYR_BASE)
         self.modules = [module.meta.get('name') for module in modules_meta]
 
+        # Check each module for a snippets directory, then add to a list if it is found
+        for module in modules_meta:
+            snippet_path = module.project + '/snippets'
+            if os.path.exists(snippet_path):
+                self.snippet_roots.append(Path(module.project))
```
We should not have a deviating implementation in twister where snippets are picked up differently than in the build system.

The build system uses the information provided in `zephyr/module.yml` but this code checks for the existence of a folder regardless of the setting in `zephyr/module.yml`.

tejlmand reviewed (changes requested):
Please look at the current support for snippet_root in modules:
https://docs.zephyrproject.org/latest/develop/modules.html#build-settings

https://github.com/zephyrproject-rtos/zephyr/blob/72ec6ce4db682c89bdcdd5688294ca075b8832a0/scripts/zephyr_module.py#L275-L276

nordicjm committed:
twister: Fix lack of support for module snippet_root

Adds support for using the Zephyr module snippet_root variable
in twister for required_snippets

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
